pkg/filesystem/subvolume.go

```diff
@@ -211,7 +217,89 @@ func Delete(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespa
 	if !check {
 		exec.RunCommandInOperatorPod(ctx, clientsets, "ceph", []string{"fs", "subvolume", "rm", fs, subvol, svg, "--retain-snapshots"}, OperatorNamespace, CephClusterNamespace, true)
 		logging.Info("subvolume %q deleted", subvol)
+		deleteOmap(ctx, clientsets, OperatorNamespace, CephClusterNamespace, subvol, fs)
 	} else {
 		logging.Info("subvolume %q is not stale", subvol)
 	}
 }
+
+// deleteOmap deletes omap object and key for the given subvolume.
+func deleteOmap(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespace, CephClusterNamespace, subVol, fs string) {
+	omapdetails := getOmapDetails(ctx, clientsets, OperatorNamespace, CephClusterNamespace, subVol, fs)
+	poolName := fs + "-metadata"
+	if omapdetails.omapval != "" {
+		// remove omap object.
+		exec.RunCommandInOperatorPod(ctx, clientsets, "rados", []string{"rm", omapdetails.omapval, "-p", poolName, "--namespace", "csi"}, OperatorNamespace, CephClusterNamespace, true)
+		logging.Info("omap object:%q deleted", omapdetails.omapval)
+	}
+	if omapdetails.omapkey != "" {
+		// remove omap key.
+		exec.RunCommandInOperatorPod(ctx, clientsets, "rados", []string{"rmomapkey", "csi.volumes.default", omapdetails.omapkey, "-p", poolName, "--namespace", "csi"}, OperatorNamespace, CephClusterNamespace, true)
+		logging.Info("omap key:%q deleted", omapdetails.omapkey)
+	}
+}
+
+// getOmapDetails gets the omap key and value details for a given subvolume.
+// Deletion of omap object required the rbd-image-name which is of format
+// csi.volume.subvolume, where subvolume is the name of subvolume that needs to be
+// deleted.
+// similarly to delete of omap key requires csi.volume.ompakey, where
+// omapkey is the pv name which is extracted the omap object.
+func getOmapDetails(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespace, CephClusterNamespace, subVol, fs string) omapDetails {
+
+	var details omapDetails
+	var omapval, pv string
+	poolName := fs + "-metadata"
+	replace := strings.Replace(subVol, "-", ".", 2)
+	checknfs := strings.Contains(subVol, "nfs-export")
+	// check if it is cephfs or nfs subvolume
+	if checknfs {
+		omapval = strings.Replace(replace, "export", "volume", 1)
+		omapval = strings.Replace(omapval, "nfs", "csi", 1)
+	} else {
+		omapval = strings.Replace(replace, "vol", "volume", 1)
+
+	}
+	ls, err := exec.RunCommandInOperatorPod(ctx, clientsets, "rados", []string{"listomapvals", omapval, "-p", poolName, "--namespace", "csi"}, OperatorNamespace, CephClusterNamespace, true)
+	if err != nil {
+		if !strings.Contains(err.Error(), "No such file or directory") {
+			logging.Fatal(fmt.Errorf("failed to list omapvals: %q", err))
+		} else {
```

Comment:
No need for "else" since the previous condition will exit after: `logging.Fatal()`
```suggestion
		}
```

Reply:
But what if the condition is not true and there are no omap details, in that case, it will execute the else statement. I have tested this condition.

Reply:
As a general pattern, it is simpler to have fewer code blocks. If we return from an `if` block, there is no need to have an `else` block. For example, this code is the same without the `else` block, because `logging.Fatal()` inside the `if` exits the program immediately.
```go
	if err != nil {
		if !strings.Contains(err.Error(), "No such file or directory") {
			// exit immediately
			logging.Fatal(fmt.Errorf("failed to list omapvals: %q", err))
		}
		logging.Info("No omapvals found for subvolume %s", subVol)
		return omapDetails{}
	}
```
But since it's equivalent in functionality, it's just a nit.